### PR TITLE
8266845: Shenandoah: Simplify SBS::load_reference_barrier implementation

### PR DESCRIPTION
**Pull request by zhengyu123**
Early return on null obj simplifies implementation and makes code more readable.

Test:
- [x] hotspot_gc_shenandoah

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266845](https://bugs.openjdk.java.net/browse/JDK-8266845): Shenandoah: Simplify SBS::load_reference_barrier implementation

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 4a70d935d42e92d9463950fc1bb6f87f00a54757

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3956/head:pull/3956` \
`$ git checkout pull/3956`

Update a local copy of the PR: \
`$ git checkout pull/3956` \
`$ git pull https://git.openjdk.java.net/jdk pull/3956/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3956`

View PR using the GUI difftool: \
`$ git pr show -t 3956`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3956.diff">https://git.openjdk.java.net/jdk/pull/3956.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp [hunk 1]
```diff
@@ -100,31 +100,35 @@ inline oop ShenandoahBarrierSet::load_reference_barrier(oop obj) {
 
 template <DecoratorSet decorators, class T>
 inline oop ShenandoahBarrierSet::load_reference_barrier(oop obj, T* load_addr) {
+  if (obj == nullptr) {
+    return nullptr;
+  }
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:105]: Let's use `NULL` consistently? Yes, it is an old style, but it is consistent with the code in the same file.

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp [hunk 2]
```diff
@@ -100,31 +100,35 @@ inline oop ShenandoahBarrierSet::load_reference_barrier(oop obj) {
...
+  if (HasDecorator<decorators, AS_NO_KEEPALIVE>::value &&
       _heap->is_evacuation_in_progress() &&
       !_heap->marking_context()->is_marked(obj)) {
     return obj;
   }
 
   oop fwd = load_reference_barrier(obj);
-  if (ShenandoahSelfFixing && load_addr != NULL && fwd != obj) {
+  assert(fwd != nullptr, "Sanity");
+  if (ShenandoahSelfFixing && load_addr != NULL) {
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:131]: I think dropping `fwd != obj` is not about the null-check. It is about not doing the CAS when evacuation fails. It is a minor performance optimization: do not try to perform obviously no-op update. Let's leave it as is.

**Review by shipilev — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp (hunk 2 above)
Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp]: Right, my mistake.

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp [hunk 3]
```diff
@@ -100,30 +100,34 @@ inline oop ShenandoahBarrierSet::load_reference_barrier(oop obj) {
...
   // concurrent class-unloading.
-  if (HasDecorator<decorators, AS_NO_KEEPALIVE>::value && obj != NULL &&
+  if (HasDecorator<decorators, AS_NO_KEEPALIVE>::value &&
       _heap->is_evacuation_in_progress() &&
       !_heap->marking_context()->is_marked(obj)) {
     return obj;
   }
 
   oop fwd = load_reference_barrier(obj);
+  assert(fwd != nullptr, "Sanity");
```
[the remark this answers is not in the document]
Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp]: No good reason. I removed it.

**Pull request closed without merging**